### ISSUE TITLE
同步上游分支将 ultralytics_yolov5 迁移到 one-yolov5 小结

### ISSUE THREAD

**Issue by ccssu** [Oneflow-Inc/one-yolov5#107]
- [同步上游分支](#同步上游分支)
  - [将 ultralytics\_yolov5 迁移到 one-yolov5 所做的改动](#将-ultralytics_yolov5-迁移到-one-yolov5-所做的改动)
  - [loss 对齐指南](#loss-对齐指南)
    - [前言](#前言)
    - [准备工作](#准备工作)
    - [train.py loss对齐](#trainpy-loss对齐)
    - [segment/train.py loss对齐](#segmenttrainpy-loss对齐)
    - [classify/train.py loss](#classifytrainpy-loss)
  - [总结和未来展望](#总结和未来展望)

#  同步上游分支
## 将 ultralytics_yolov5 迁移到 one-yolov5 所做的改动

```python
import torchvision.  => import flowvision.
import torchvision  => import flowvision as torchvision
form torch. => from oneflow. 
import torch. => import oneflow.
import torch => import oneflow as torch 
```
在OneFlow 发布 v0.9.0 版本 后，主要的修改在于头文件的替换，如果感兴趣看看从torch迁移到oneflow都做了哪些改动，可以查看这个链接：https://github.com/Oneflow-Inc/one-yolov5/compare/ultralytics_yolov5_064365d8683fd002e9ad789c1e91fa3d021b44f0...Synchronize_ultralytics_yolov5_064365d8683fd002e9ad789c1e91fa3d021b44f0 

### 前言 
1. 加载同一个预训练权重。
2. shuffle设置为False，保证加载的数据一致
3. 使用FP32，fp16训练研发中，相信不久将会上线。
4. OneFlow版本 >= 0.9.0

### 准备工作

1. 使用代码
    - OneFlow后端代码：
    ```shell
    $ git clone git@github.com:Oneflow-Inc/one-yolov5.git && git checkout debug_loss_Synchronize_ultralytics_yolov5_064365d8683fd002e9ad789c1e91fa3d021b44f0
    ```
    - PyTorch后端代码：
    ```shell
    $ git clone git@github.com:Oneflow-Inc/one-yolov5.git && git checkout debug_loss_ultralytics_yolov5_064365d8683fd002e9ad789c1e91fa3d021b44f0
    ```
2. `pip install -r requirements.txt`
3. 须要的预训练模型：
    - https://github.com/ultralytics/yolov5/releases/download/v7.0/yolov5s-seg.pt
    - https://github.com/ultralytics/yolov5/releases/download/v7.0/yolov5s.pt
4. 需要的数据集：
    - https://oneflow-static.oss-cn-beijing.aliyuncs.com/one-yolo/coco128-seg.zip
    - https://oneflow-static.oss-cn-beijing.aliyuncs.com/one-yolo/coco128.zip

5. 修改项目中 `data/` 路径下修改为对应下载数据集的路径。

注意：
    使用OneFlow版本 >= 0.9.0 , PyTorch >= 1.13.0

### train.py loss对齐 

- 启动指令: 
```shell
$ python train.py --data coco128.yaml --weights ../datasets/classify/weights/yolov5s.pt  --img 640 --epochs 300 --noval
```
注意替换权重文件路径

**下图 `train.py`  前 300 个 epochs 对齐情况  (使用 `optimizer=SGD`)**
![image](https://user-images.githubusercontent.com/109639975/212058633-0387e68d-c4d1-480d-847d-7a3c7a48cf24.png)
图1.  epochs = 300 optimizer=SGD  loss对齐情况。

### segment/train.py loss对齐
- 启动指令:
```shell
$ python segment/train.py --data coco128-seg.yaml --weights ../datasets/yolov5s-seg.pt --img 640 --epochs 300 --noval
```
注意替换权重文件路径

**下图 `segment/train.py`  前 300 个 epochs 对齐情况  (使用 `optimizer=SGD`)**
![300个epoch的loss对齐情况](https://foruda.gitee.com/images/1673969811319575402/af45b656_10213136.png "")
图2.  epochs = 300 optimizer=SGD  loss对齐情况。

### classify/train.py loss 
这里暂时不验证 `classify/train.py loss ` ， 
原因：
1. oneflow 不能直接加载torch的module对象，是oneflow 的module直接加载pytorch权重。
2. 尝试直接加载torch模型后直接转换为oneflow模型，出现模型层数不一致，具体是 BatchNorm2d 后缺失 Silu 激活函数。
3. 如下图所示，通过加载同一模型，np.allclose() 验证` Classify` 前面层数测试 测试输出相等，可以是由 (1,2) 产生的差异导致 `Classify` 权重加载有点差异，尝试先试下验证精度，精度达标忽略验证 `classify/train.py loss`
![image](https://user-images.githubusercontent.com/109639975/216230127-8ba52f07-1d19-40fe-81af-df68fbd981cf.png)

## 总结和未来展望
本文介绍了使用OneFlow 将 ultralytics_yolov5 迁移到 one-yolov5 所做的改动，并进行loss对齐测试。
未来：模型导出和hub功能修复。 目标精度以及显存，速度测试。
